src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -189,40 +207,129 @@ irep_idt generate_java_generic_typet::build_generic_tag(
-  new_tag_buffer << "<";
-  bool first=true;
-  for(const typet &type_argument : existing_generic_type
-    .generic_type_arguments())
+  const std::string &original_class_tag = original_class.get_tag().c_str();
+  auto generic_argument_p =
+    existing_generic_type.generic_type_arguments().begin();
+
+  // if the original class is implicitly generic, add type
+  // arguments for all generic outer classes
+  if(is_java_implicitly_generic_class_type(original_class))
   {
-    if(!first)
-      new_tag_buffer << ", ";
-    first=false;
-
-    INVARIANT(
-      !is_java_generic_parameter(type_argument),
-      "Only create full concretized generic types");
-    const irep_idt &id(id2string(type_argument.subtype().get(ID_identifier)));
-    new_tag_buffer << pretty_print_java_type(id2string(id));
-    if(is_java_array_tag(id))
+    const java_implicitly_generic_class_typet
+      &implicitly_generic_original_class =
+        to_java_implicitly_generic_class_type(original_class);
+    const auto &implicit_generic_types =
+      implicitly_generic_original_class.implicit_generic_types();
+
+    std::string current_outer_class_name =
+      original_class_tag.substr(0, original_class_tag.find("$"));
+    new_tag_buffer << current_outer_class_name;
+    bool first_outer = true;
+    bool first = true;
+
+    for(const auto &parameter : implicit_generic_types)
     {
-      const typet &element_type =
-        java_array_element_type(to_symbol_type(type_argument.subtype()));
-
-      // If this is an array of references then we will specialize its
-      // identifier using the type of the objects in the array. Else, there can
-      // be a problem with the same symbols for different instantiations using
-      // arrays with different types.
-      if(element_type.id() == ID_pointer)
+      PRECONDITION(is_java_generic_parameter(parameter));
+      const std::string &parameter_name = to_java_generic_parameter(parameter)
+                                            .type_variable()
+                                            .get_identifier()
+                                            .c_str();
+      // get the class name without the `java::` prefix, e.g., for
+      // `java::Class::T` parameter_class_name is `Class`
+      const std::string &parameter_class_name =
+        parameter_name.substr(6, parameter_name.rfind("::") - 6);
+
+      if(parameter_class_name != current_outer_class_name)
       {
-        const symbol_typet element_symbol =
-          to_symbol_type(element_type.subtype());
-        new_tag_buffer << "of_" << id2string(element_symbol.get_identifier());
+        // we move on to the next class, which is an inner class of the
+        // current_outer_class_name, i.e., current_outer_class_name is a
+        // prefix of parameter_class_name
+        PRECONDITION(
+          parameter_class_name.compare(current_outer_class_name) > 0);
+        if(!first_outer)
+          new_tag_buffer << ">";
```

Comment:
I agree with @smowton that this has got a bit long - you might consider pulling out a little class for building up the angle brackets of a specific class that can track whether it has had any parameters put into it (you'd have a "to_string" method that would return the empty string if no parameters added at that level or something).

This would also reduce the number of variables to keep track over the loop and make it easier in future to make the for loop not require the parameters to be in a specific order.

Reply:
And I suppose would allow some de-duplication with the case it itself is a java_generic_class

Reply:
The function has been divided into a number of smaller functions now.